net/graphsync_fetcher.go

```
@@ -78,38 +88,79 @@ const recursionMultiplier = 4
+			break
+		}
+		// something went wrong in a graphsync request, but we want to keep trying other peers, so
+		// just log error
+		logGraphsyncFetcher.Infof("Error occurred during Graphsync request: %s", err)
+		cids, err = gsf.missingCids(cids)
+		if err != nil {
+			return types.UndefTipSet, err
+		}
+		err = rpf.FindNextPeer()
+		if err != nil {
+			return types.UndefTipSet, fmt.Errorf("Failed fetching tipset: %s", tsKey.String())
+		}
 	}
+	return gsf.loadTipsetFromCids(ctx, tsKey.ToSlice())
+}
+
+func (gsf *GraphSyncFetcher) fetchRemainingTipsets(ctx context.Context, startingTipset types.TipSet, rpf *requestPeerFinder, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
 	out := []types.TipSet{startingTipset}
 	isDone, err := done(startingTipset)
 	if err != nil {
 		return nil, err
 	}
+
+	// fetch remaining tipsets recursively
 	recursionDepth := 1
 	ts := startingTipset
 	for !isDone {
 		// Because a graphsync query always starts from a single CID,
 		// we fetch tipsets starting from the first block in the last tipset and
 		// recursively getting sets of parents
-		err := gsf.fetchBlocksRecursively(ctx, ts.At(0).Cid(), from, recursionDepth)
+		err := gsf.fetchBlocksRecursively(ctx, ts.At(0).Cid(), rpf.CurrentPeer(), recursionDepth)
 		if err != nil {
-			return nil, err
+			// something went wrong in a graphsync request, but we want to keep trying other peers, so
+			// just log error
+			logGraphsyncFetcher.Infof("Error occurred during Graphsync request: %s", err)
 		}
+		cidsAreMissing := false
 		for i := 0; i < recursionDepth; i++ {
 			tsKey, err := ts.Parents()
 			if err != nil {
 				return nil, err
 			}
-			cids := tsKey.ToSlice()
-			ts, err = gsf.loadTipsetFromCids(ctx, cids)
+			cidsAreMissing, err = gsf.areCidsMissing(tsKey)
+			if err != nil {
+				return nil, err
+			}
+			if cidsAreMissing {
+				err := rpf.FindNextPeer()
+				if err != nil {
+					return nil, fmt.Errorf("Failed fetching tipset: %s", tsKey.String())
```

Ditto swallowed error